File: playground/android/app/src/main/java/com/amap/agenuiplayground/PlaygroundRuntimeLogger.java
package com.amap.agenuiplayground;

import android.util.Log;

import com.amap.agenui.IAGenUILogger;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

/**
 * Custom RuntimeLogger implementation for AGenUI Playground
 * 
 * This logger forwards AGenUI engine logs to Android Logcat with proper formatting.
 * It also provides a callback interface for displaying logs in the UI.
 */
public class PlaygroundRuntimeLogger implements IAGenUILogger {
    private static final String TAG = "AGenUI-Engine";
    
    public PlaygroundRuntimeLogger() {
        // Default constructor
    }
    
    @Override
    public void onLog(int level, String tag, String func, int line, String message) {
        // Get current timestamp
        String timestamp = getTimestamp();
        
        // Format the log message with timestamp
        String formattedMessage = String.format("%s [%s@%d] %s", timestamp, func, line, message);
        String logTag = (tag != null && !tag.isEmpty()) ? tag : TAG;

        // Always log to Android Logcat based on level
        switch (level) {
            case 0: // DEBUG
                Log.d(logTag, formattedMessage);
                break;
            case 1: // INFO
                Log.i(logTag, formattedMessage);
                break;
            case 2: // WARN
                Log.w(logTag, formattedMessage);
                break;
            case 3: // ERROR
                Log.e(logTag, formattedMessage);
                break;
            case 4: // FATAL
                Log.wtf(logTag, formattedMessage);
                break;
            case 5: // PERFORMANCE
                Log.d(logTag, "[PERF] " + formattedMessage);
                break;
            default:
                Log.d(logTag, formattedMessage);
                break;
        }
    }
    
    /**
     * Get human-readable log level name
     */
    public static String getLevelName(int level) {
        switch (level) {
            case 0: return "DEBUG";
            case 1: return "INFO";
            case 2: return "WARN";
            case 3: return "ERROR";
            case 4: return "FATAL";
            case 5: return "PERF";
            default: return "UNKNOWN";
        }
    }
    
    /**
     * Get formatted timestamp for log messages
     * Format: HH:mm:ss.SSS (e.g., 14:30:25.123)
     */
    private String getTimestamp() {
        SimpleDateFormat sdf = new SimpleDateFormat("HH:mm:ss.SSS", Locale.getDefault());
        return sdf.format(new Date());
    }
}

File: playground/ios/Playground/Playground for AGenUI/Utils/PlaygroundRuntimeLogger.swift
//
//  PlaygroundRuntimeLogger.swift
//  Playground
//
//  Created by yinglong.zyl on 2026/5/8.
//

import Foundation
import AGenUI

/// Playground custom RuntimeLogger
///
/// Implements the LoggerDelegate protocol, receives runtime logs from the AGenUI C++ engine,
/// and formats output to the Xcode console for development debugging.
///
/// Usage: Call AGenUISDK.setLoggerDelegate(PlaygroundRuntimeLogger.shared) in AppDelegate
final class PlaygroundRuntimeLogger: NSObject, LoggerDelegate {

    static let shared = PlaygroundRuntimeLogger()

    private override init() {
        dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "HH:mm:ss.SSS"
    }

    private let dateFormatter: DateFormatter

    // MARK: - LoggerDelegate

    func onLog(level: Logger.Level, tag: String, func: String, line: Int, message: String) {
        let prefix = levelPrefix(for: level)
        let timestamp = dateFormatter.string(from: Date())
        print("[AGenUI/\(prefix)] \(timestamp) [\(tag):\(line)] \(message)")
    }

    // MARK: - Private

    private func levelPrefix(for level: Logger.Level) -> String {
        switch level {
        case .debug:       return "DEBUG"
        case .info:        return "INFO"
        case .warning:     return "WARN"
        case .error:       return "ERROR"
        case .fatal:       return "FATAL"
        case .performance: return "PERF"
        }
    }
}
